README.md
# Typesetter's Son

[![npm](https://img.shields.io/npm/v/typesetters-son.svg?style=flat-square)](https://www.npmjs.com/package/typesetters-son)
[![license](https://img.shields.io/npm/l/typesetters-son.svg?style=flat-square)](https://github.com/chromakode/typesetters-son/blob/master/LICENSE)

[![Named for "Typesetter's Son" on Channel101](typesetters-son.jpg)](http://www.channel101.com/episode/1667)

Typesetter's Son is a small utility for generating images and PDFs based on lightly modified webpages. It loads a URL in [Headless Chrome](https://github.com/GoogleChrome/puppeteer), substitutes innerHTML for elements you specify, and takes a screenshot.

It works great on SVGs too -- just sprinkle in `id` attributes where you want to modify text!

## Installation

Library: `npm install typesetters-son`  
Executable: `npm install -g typesetters-son`

## Example:

`typesetters-son http://dogpatchjs.com/banner -o dogpatch.png -w 1024 -c 'body' -s '{"#day": "Monday, July 17, 2017", "#where": "Spark Social", "#time": "6:30pm"}'`

`typesetters-son http://dogpatchjs.com/banner -o dogpatch.pdf -p Letter -s '{"#day": "Monday, July 17, 2017", "#where": "Spark Social", "#time": "6:30pm"}'`

![Example output](example.png)

## Library Usage

```js
const {generateImage} = require('typesetters-son')
generateImage({
  url: 'http://dogpatchjs.com/banner',
  output: 'dogpatch.png',
  width: 1024,
  height: 800,
  density: 1,  // defaults to 1 if unspecified
  crop: 'body',
  subs: {
   '#day': new Date().toDateString(),
   '#where': 'Spark Social',
   '#time': '6:30pm',
  },
})

const {generatePDF} = require('typesetters-son')
generatePDF({
  url: 'http://dogpatchjs.com/banner',
  output: 'dogpatch.pdf',
  paper: 'Letter',
  removeBackground: false,
  subs: {
   '#day': new Date().toDateString(),
   '#where': 'Spark Social',
   '#time': '6:30pm',
  },
})
```

## Executable Usage

```
Usage: typesetters-son URL -o [path]

Options:
  --output, -o   output image/pdf path                                [required]
  --width, -w    viewport width                                   [default: 800]
  --height, -h   viewport height                                  [default: 600]
  --crop, -c     element query to crop around
  --density, -d  pixel density                                      [default: 1]
  --paper, -p    paper size (e.g. Letter, A4, etc.)          [default: "Letter"]
  --remove-bg    remove bg colors (for printing)      [boolean] [default: false]
  --subs, -s     text substitutions (JSON)
```


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## 2.0.0 - 2019-01-25
### Added
- PDF generation.
- Updated `puppeteer`.

### Changed
- Now requires Node > 8.
- Module API now exports a {generateImage, generatePDF} object.


## 1.3.1 - 2017-12-02
### Fixed
- Fixed use of removed 'networkidle' waitUntil option.

## 1.3.0 - 2017-12-02
### Added
- `browserWSEndpoint` parameter for connecting to a running Chrome.
- Updated `puppeteer`.

## 1.2.0 - 2017-09-04
### Added
- Node 6 support.

## 1.1.0 - 2017-09-03
### Added
- Importable library API.

## 1.0.0 - 2017-09-01
### Added
- Initial public release.
- Executable utility.


src/index.js
const puppeteer = require('puppeteer')

async function performSubstitutions(page, subs) {
  await page.evaluate(subs => {
    for (const query of Object.keys(subs)) {
      const text = subs[query]
      document.querySelectorAll(query).forEach(el => {
        el.innerHTML = text
      })
    }
  }, subs)
}

async function generateImage({
  browserWSEndpoint,
  url,
  output,
  width,
  height,
  density=1,
  subs,
  crop,
}) {
  let browser
  if (browserWSEndpoint) {
    browser = await puppeteer.connect({browserWSEndpoint})
  } else {
    browser = await puppeteer.launch()
  }
  try {
    const page = await browser.newPage()
    await page.setViewport({
      width: width,
      height: height,
      deviceScaleFactor: density,
    })

    await page.goto(url, {waitUntil: 'networkidle0'})

    if (subs) {
      await performSubstitutions(page, subs)
    }

    let clip
    if (crop) {
      clip = await page.evaluate(query => {
        const el = document.querySelector(query)
        if (!el) {
          return
        }
        const box = el.getBoundingClientRect()
        return {
          x: box.left,
          y: box.top,
          width: box.width,
          height: box.height,
        }
      }, crop)
    }

    return await page.screenshot({path: output, clip})
  } finally {
    browser.close()
  }
}

async function generatePDF({
  browserWSEndpoint,
  url,
  output,
  subs,
  paper,
  removeBackground=false,
}) {
  let browser
  if (browserWSEndpoint) {
    browser = await puppeteer.connect({browserWSEndpoint})
  } else {
    browser = await puppeteer.launch()
  }
  try {
    const page = await browser.newPage()
    await page.goto(url, {waitUntil: 'networkidle0'})

    if (subs) {
      await performSubstitutions(page, subs)
    }

    return await page.pdf({
      path: output,
      format: paper,
      printBackground: !removeBackground,
    })
  } finally {
    browser.close()
  }
}

module.exports = {
  generateImage,
  generatePDF,
}


src/bin.js
#!/usr/bin/env node
const {generateImage, generatePDF} = require('./')

const argv = require('yargs')
  .usage('Usage: $0 URL -o [path]')
  .option('output', {
    alias: 'o',
    describe: 'output image/pdf path',
  })
  .option('width', {
    alias: 'w',
    describe: 'viewport width',
    default: 800,
  })
  .option('height', {
    alias: 'h',
    describe: 'viewport height',
    default: 600,
  })
  .option('crop', {
    alias: 'c',
    describe: 'element query to crop around',
  })
  .option('density', {
    alias: 'd',
    describe: 'pixel density',
    default: 1,
  })
  .option('paper', {
    alias: 'p',
    describe: 'paper size (e.g. Letter, A4, etc.)',
    default: 'Letter',
  })
  .option('remove-bg', {
    describe: 'remove bg colors (for printing)',
    default: false,
    type: 'boolean',
  })
  .option('subs', {
    alias: 's',
    describe: 'text substitutions (JSON)',
    coerce: JSON.parse,
  })
  .demand('o')
  .demandCommand(1, 'Please specify a URL')
  .argv

argv.url = argv._[0]
const generate = argv.output.endsWith('.pdf')
  ? generatePDF
  : generateImage

generate(argv)
  .catch(err => {
    console.log(err.toString())
  })
